src/routing/bootstrap.rs

```diff
@@ -336,70 +173,125 @@ impl<'a> State<'a> {
+        relocate_details: Option<SignedRelocateDetails>,
     ) -> Result<(Node, Section, Vec<(RoutingMsg, SocketAddr, DestInfo)>)> {
         let join_request = JoinRequest {
             section_key,
-            relocate_payload: relocate_payload.clone(),
+            relocate_payload: None,
             resource_proof_response: None,
         };
-        let recipients = elders
-            .into_iter()
-            .map(|(name, addr)| (name, addr))
-            .collect_vec();
-        self.send_join_requests(join_request, recipients, section_key)
+
+        // Avoid sending more than one request to the same peer.
+        let mut used_recipient = HashSet::<(XorName, SocketAddr)>::new();
+
+        self.send_join_requests(join_request, &recipients, section_key)
             .await?;
 
+        let mut relocate_payload = None;
         loop {
+            used_recipient.extend(recipients.iter());
+
             let (response, sender, dest_info) = self
                 .receive_join_response(genesis_key.as_ref(), relocate_payload.as_ref())
                 .await?;
 
             match response {
+                JoinResponse::Rejected(JoinRejectionReason::NodeNotReachable(addr)) => {
+                    error!(
+                        "Node cannot join the network since it is not externally reachable: {}",
+                        addr
+                    );
+                    return Err(Error::NodeNotReachable(addr));
+                }
+                JoinResponse::Rejected(JoinRejectionReason::JoinsDisallowed) => {
+                    error!("Network is set to not taking any new joining node, try join later.");
+                    return Err(Error::TryJoinLater);
+                }
                 JoinResponse::Approval {
                     section_auth,
                     genesis_key,
                     section_chain,
+                    ..
                 } => {
                     return Ok((
                         self.node,
                         Section::new(genesis_key, section_chain, section_auth)?,
                         self.backlog.into_iter().collect(),
                     ));
                 }
-                JoinResponse::Retry {
-                    section_auth,
-                    section_key: new_section_key,
-                } => {
-                    if new_section_key == section_key {
+                JoinResponse::Retry(section_auth) | JoinResponse::Redirect(section_auth) => {
+                    if section_auth.section_key() == section_key {
                         continue;
                     }
 
-                    if section_auth.prefix().matches(&self.node.name()) {
+                    // Ignore already used recipients
+                    let new_recipients: Vec<(XorName, SocketAddr)> = section_auth
+                        .elders
+                        .iter()
+                        .filter(|(name, addr)| !used_recipient.contains(&(**name, **addr)))
```

Comment:
1, for the case of `retry` you may need to contact the same guy again ?
2, only need to check name or address? why need to check the pair?

Reply:
good catch @maqi , thanks, I'll fix 1).
For 2), I guess I can check the address only.